common/src/main/java/com/viaversion/viaversion/connection/ConnectionManagerImpl.java

```diff
@@ -35,24 +35,25 @@
     protected final Set<UserConnection> connections = Collections.newSetFromMap(new ConcurrentHashMap<>());
 
     @Override
-    public void onLoginSuccess(UserConnection connection) {
+    public synchronized void onLoginSuccess(UserConnection connection) {
         Objects.requireNonNull(connection, "connection is null!");
-        connections.add(connection);
+        boolean newlyAdded = connections.add(connection);
 
         if (isFrontEnd(connection)) {
             UUID id = connection.getProtocolInfo().getUuid();
-            if (clients.put(id, connection) != null) {
+            UserConnection previous = clients.put(id, connection);
+            if (previous != null && previous != connection) {
                 Via.getPlatform().getLogger().warning("Duplicate UUID on frontend connection! (" + id + ")");
             }
         }
 
-        if (connection.getChannel() != null) {
+        if (newlyAdded && connection.getChannel() != null) {
             connection.getChannel().closeFuture().addListener((ChannelFutureListener) future -> onDisconnect(connection));
         }
     }
 
     @Override
-    public void onDisconnect(UserConnection connection) {
+    public synchronized void onDisconnect(UserConnection connection) {
```

Comment:
No need to synchronize the onDisconnect at all

Reply:
Without synchronizing onDisconnect, onLoginSuccess can be adding at-the-same-time as disconnect is removing, where onLoginSuccess tries to connections.add which returns false, as in, connection is already there, and onDisconnect removes it instants later

Then onLoginSuccess proceeds to add it to the map, (where it will be left forever) and a closeFuture will not be dispatched, as "newlyAdded" was false